File: cpp/bst.cpp
#include <iostream>

template <typename T>
class _bstreenode;

template <typename T>
class BSTree;

template <typename T>
class _bstreenode {
 protected:
  T data;
  _bstreenode *left, *right;

 public:
  _bstreenode() {}
  _bstreenode(T data) : data(data), left(nullptr), right(nullptr) {}
  _bstreenode(_bstreenode* left, T data)
      : data(data), left(left), right(nullptr) {}
  _bstreenode(T data, _bstreenode* right)
      : data(data), left(nullptr), right(right) {}
  _bstreenode(_bstreenode* left, T data, _bstreenode* right)
      : data(data), left(left), right(right) {}

  friend class BSTree<T>;
};

template <typename T>
class BSTree {
 protected:
  _bstreenode<T>* root;

  _bstreenode<T>*cur, *prev;
  bool dataExists;
  long size;

  void _ptrReset() {
    cur = root;
    prev = root;
  }

  void _searchBSTree(T data) {
    if (cur == nullptr) {
      cur = prev;
      return;
    }

    if (data == cur->data) {
      this->dataExists = true;
      return;
    }

    prev = cur;
    if (data > cur->data) {
      cur = cur->right;
    } else if (data < cur->data) {
      cur = cur->left;
    }

    _searchBSTree(data);

    return;
  }

  void _inorderSucc() {
    prev = cur;
    cur = cur->right;
    while (cur != nullptr && cur->left != nullptr) {
      prev = cur;
      cur = cur->left;
    }
  }

  void _inorder() {
    if (this->cur == nullptr) return;

    _bstreenode<T>* cur = this->cur;

    this->cur = cur->left;
    _inorder();

    std::cout << cur->data << " ";

    this->cur = cur->right;
    _inorder();
  }

  void _removeHelper(T data) {
    _searchBSTree(data);

    if (!dataExists) {
      return;
    }

    dataExists = false;

    // root node with no succ:
    if (prev == cur && cur->right == nullptr) {
      root = root->left;
      prev = root;

      this->size--;
      delete cur;
      cur = nullptr;
      return;
    }

    // both child
    if (cur->left != nullptr && cur->right != nullptr) {
      _bstreenode<T>* el = cur;
      _inorderSucc();
      el->data = cur->data;
      this->cur = el->right;
      _removeHelper(el->data);

      return;
    }

    // leaf node or @least 1 child;
    _bstreenode<T>* pointTo = (cur->right != nullptr) ? cur->right : cur->left;

    if (cur == prev->right)
      prev->right = pointTo;
    else if (cur == prev->left)
      prev->left = pointTo;
    else  // root node
      root = pointTo;

    this->size--;
    delete cur;
    cur = nullptr;
    return;
  }

  void _clear() {
    if (this->cur == nullptr) return;
    _bstreenode<T>* cur = this->cur;

    this->cur = cur->left;
    _clear();

    this->cur = cur->right;
    _clear();

    delete cur;
    cur = nullptr;
  }

 public:
  BSTree() : root(nullptr), cur(this->root), size(0), dataExists(false) {}

  ~BSTree() {
    _ptrReset();
    _clear();
    root = nullptr;
  }

  BSTree<T>& insert(T data) {
    _ptrReset();
    _searchBSTree(data);

    // condition when cur = root;
    if (cur == nullptr) {
      root = new _bstreenode<T>(data);
      return *this;
    }

    _bstreenode<T>* newNode = new _bstreenode<T>(data);
    if (data > cur->data) {
      cur->right = newNode;
    } else {
      cur->left = newNode;
    }
    this->size++;

    return *this;
  }

  BSTree<T>& remove(T data) {
    _ptrReset();
    _removeHelper(data);

    return *this;
  }

  BSTree<T>& clear() {
    _ptrReset();
    _clear();

    this->root = nullptr;
    this->size = 0;

    return *this;
  }

  bool search(T data) {
    _ptrReset();
    _searchBSTree(data);

    bool found = dataExists;
    dataExists = false;

    return found;
  }

  void inorder() {
    _ptrReset();
    _inorder();
    std::cout << "\n";
  }
};

int main() { return 0; }

File: typescript/bst.ts
class _bstreenode {
  public data: number;
  public left: _bstreenode | null;
  public right: _bstreenode | null;

  constructor(d: number);
  constructor(d: number) {
    this.data = d;
    this.left = null;
    this.right = null;
  }
}

class BSTree {
  private root: _bstreenode | null;
  private cur: _bstreenode | null;
  private prev: _bstreenode | null;
  private size: number;

  private dataExists: boolean;

  private _ptrReset() {
    this.cur = this.root;
    this.prev = this.root;
    this.dataExists = false;
  }

  private _bstreeSearch(data: number) {
    if (this.cur === null) {
      this.cur = this.prev;
      return false;
    }

    if (data === this.cur?.data) {
      this.dataExists = true;
      return true;
    }

    this.prev = this.cur;
    if (data < this.cur?.data) {
      this.cur = this.cur.left;
    } else {
      this.cur = this.cur?.right;
    }

    return this._bstreeSearch(data);
  }

  private _inorderSucc() {
    if (this.cur === null) return;
    this.cur = this.cur.right;

    while (this.cur != null && this.cur.left != null) {
      this.prev = this.cur;
      this.cur = this.cur.left;
    }

    return;
  }

  private _removeHelper(data: number) {
    this._bstreeSearch(data);

    if (!this.dataExists || this.cur === null || this.prev == null) {
      return;
    }

    this.dataExists = false;
    this.size--;

    // root node with no inorder successor
    if (this.cur == this.prev && this.cur?.right === null) {
      this.root = this.cur?.left;
      // Garbage Collector in working
      return;
    }

    // both the children exist
    if (this.cur?.left != null && this.cur?.right != null) {
      const el: _bstreenode = this.cur;
      this._inorderSucc();
      el.data = this.cur.data;

      this.prev = this.cur;
      this.cur = el.right;
      this.size++;
      this._removeHelper(el.data);

      return
    }

    // leaf node or @least 1 child
    let pointTo: _bstreenode | null =
      (this.cur.right != null)
        ? this.cur.right
        : this.cur.left;

    if (this.cur === this.prev.left) {
      this.prev.left = pointTo;
    } else if (this.cur === this.prev.right) {
      this.prev.right = pointTo
    } else {  // rot node
      this.root = pointTo;
    }

    return;
  }

  private _inorder() {
    if (this.cur === null) return;
    const cur: _bstreenode = this.cur;

    this.cur = cur.left;
    this._inorder();

    process.stdout.write(`${cur.data} `);

    this.cur = cur.right;
    this._inorder();

    return;
  }

  constructor() {
    this.root = null;
    this.cur = this.root;
    this.prev = this.root;
  }

  public search(data: number): boolean {
    this._ptrReset();
    return this._bstreeSearch(data);
  }

  public insert(data: number): void {
    this._ptrReset();
    this._bstreeSearch(data);

    if (this.dataExists) {
      this.dataExists = false;
      return;
    }

    this.size++;

    // root condition
    if (this.root === null || this.cur === null) {
      this.root = new _bstreenode(data);
      return;
    }

    const newNode: _bstreenode = new _bstreenode(data);
    if (data < this.cur.data) {
      this.cur.left = newNode;
    } else {
      this.cur.right = newNode;
    }
  }

  public remove(data: number) {
    this._ptrReset();
    this._removeHelper(data);

    return;
  }

  public inorder() {
    this._ptrReset();
    this._inorder();

    console.log();
    return;
  }

}

const myTree = new BSTree();
const a: number[] = [4, 7, 5, 3, 2, 1]

a.forEach((el: number) => {
  myTree.insert(el);
  myTree.inorder();
})

a.forEach((el: number) => {
  myTree.remove(el);
  myTree.inorder();
})
